set.c

```diff
@@ -0,0 +1,3555 @@
+
+typedef char set_check_for_sizeof_set_index_t[SIZEOF_VOIDP == (int)sizeof(set_index_t) ? 1 : -1];
+#define SIZEOF_SET_INDEX_T SIZEOF_VOIDP
+
+struct set_hash_type {
+    int (*compare)(set_data_t, set_data_t); /* set_compare_func* */
+    set_index_t (*hash)(set_data_t);        /* set_hash_func* */
+};
+
+#define SET_INDEX_BITS (SIZEOF_SET_INDEX_T * CHAR_BIT)
+
+#if defined(HAVE_BUILTIN___BUILTIN_CHOOSE_EXPR) && defined(HAVE_BUILTIN___BUILTIN_TYPES_COMPATIBLE_P)
+# define SET_DATA_COMPATIBLE_P(type) \
+   __builtin_choose_expr(__builtin_types_compatible_p(type, set_data_t), 1, 0)
+#else
+# define SET_DATA_COMPATIBLE_P(type) 0
+#endif
+
+typedef struct set_table_entry set_table_entry;
+
+struct set_table_entry; /* defined in st.c */
+
+struct set_table {
+    /* Cached features of the table -- see st.c for more details.  */
+    unsigned char entry_power, bin_power, size_ind;
+    /* How many times the table was rebuilt.  */
+    unsigned int rebuilds_num;
+    const struct set_hash_type *type;
+    /* Number of entries currently in the table.  */
+    set_index_t num_entries;
+    /* Array of bins used for access by keys.  */
+    set_index_t *bins;
+    /* Start and bound index of entries in array entries.
+       entries_starts and entries_bound are in interval
+       [0,allocated_entries].  */
+    set_index_t entries_start, entries_bound;
+    /* Array of size 2^entry_power.  */
+    set_table_entry *entries;
+};
+
+enum set_retval {SET_CONTINUE, SET_STOP, SET_DELETE, SET_CHECK};
+
+static size_t set_table_size(const struct set_table *tbl);
+static set_table *set_init_table_with_size(const struct set_hash_type *, set_index_t);
+static int set_delete(set_table *, set_data_t *); /* returns 0:notfound 1:deleted */
+static int set_insert(set_table *, set_data_t);
+static int set_lookup(set_table *, set_data_t);
+/* *key may be altered, but must equal to the old key, i.e., the
+ * results of hash() are same and compare() returns 0, otherwise the
+ * behavior is undefined */
+typedef int set_foreach_callback_func(set_data_t, set_data_t);
+typedef int set_foreach_check_callback_func(set_data_t, set_data_t, int);
+static int set_foreach(set_table *, set_foreach_callback_func *, set_data_t);
+static int set_foreach_check(set_table *, set_foreach_check_callback_func *, set_data_t, set_data_t);
+static set_index_t set_keys(set_table *table, set_data_t *keys, set_index_t size);
+static void set_free_table(set_table *);
+static void set_clear(set_table *);
+static set_table *set_copy(set_table *);
+static CONSTFUNC(int set_numcmp(set_data_t, set_data_t));
+static PUREFUNC(size_t set_memsize(const set_table *));
+static PUREFUNC(set_index_t set_hash(const void *ptr, size_t len, set_index_t h));
+static CONSTFUNC(set_index_t set_hash_end(set_index_t h));
+static CONSTFUNC(set_index_t set_hash_start(set_index_t h));
+
+/* The type of hashes.  */
+typedef set_index_t set_hash_t;
+
+struct set_table_entry {
+    set_hash_t hash;
+    set_data_t key;
+};
```

Comment:
Is it really worth having `set_hash_t / set_data_t / set_index_t`? I might be off, but I think re-using `st_*` types would be easier and would make it trivial to covert code using an `st_table` into using a `set_table`?

Reply:
I made the types deliberately different to avoid ambiguity. However, there is no technical reason we couldn't use the same types, and I agree that it could make conversion from `st_table` easier.

If we want to reuse `st_*` types, one option is to move the `set_table` code into `st.c`.  There are some functions that are the same between the two implementations, so it would eliminate some duplicate code.  What are your thoughts on that?

Reply:
> What are your thoughts on that?

That would be fine by me.

For the record there's an in-flight PR that would also benefit from a `set_table`: https://github.com/ruby/ruby/pull/13077

Reply:
Giving some more thought to this, before making changes, I think we should discuss this at the developer meeting next week, as other developers may have thoughts about where the code should go.  I'll add this to the meeting agenda.